src/state/slices/portfolioSlice/utils/index.ts

```diff
@@ -399,3 +407,141 @@ export const haveSameElements = <T>(arr1: T[], arr2: T[]) => {
+                  case 'WBNB':
+                    return 'BNB'
+                  case 'WMATIC':
+                    return 'MATIC'
+                  case 'WAVAX':
+                    return 'AVAX'
+                  default:
+                    return asset.symbol
+                }
+              }) ?? []
+
+            // Our best effort to contruct sane name using the native asset -> asset naming hack failed, but thankfully, upstream name is very close e.g
+            // for "UniswapV2 LP TRUST/WETH", we just have to append "Pool" to that and we're gucci
+            if (assetSymbols.some(symbol => !symbol)) return `${token.name} Pool`
+            return `${platform.name} ${assetSymbols.join('/')} Pool`
+          })()
+
+          const [, ...assetImages] = maybePortalsAsset.images ?? []
+
+          const { icon, icons } = ((): Pick<Asset, 'icon' | 'icons'> => {
+            // There are no underlying tokens' images, return asset icon if it exists
+            if (!assetImages?.length) {
+              return { icon: state.assets.byId[token.assetId]?.icon }
+            }
+
+            if (assetImages.length === 1) {
+              return { icon: maybeTokenImage(maybePortalsAsset.image || assetImages[0]) }
+            }
+
+            // This is a multiple assets pool, populate icons array
+            if (assetImages.length > 1) {
+              return {
+                icons: assetImages.map((image, i) => {
+                  const token = maybePortalsAsset.tokens[i]
+
+                  // No token at that index, but this isn't reliable as we've found out, it may be missing in tokens but present in images
+                  // However, this has to be an early return and we can't use our own flavour of that asset... because we have no idea which asset it is.
+                  if (!token) return maybeTokenImage(image) || ''
+
+                  const assetId = toAssetId({ chainId, assetNamespace, assetReference: token })
+                  const asset = state.assets.byId[assetId]
+
+                  // Prioritise our own flavour of icons for that asset if available, else use upstream if present
+                  return asset?.icon || maybeTokenImage(image) || ''
+                }),
+                icon: undefined,
+              }
+            }
+
+            return { icon: undefined, icons: undefined }
+          })()
+
+          minimalAsset.name = name
+          minimalAsset.isPool = isPool
+          minimalAsset.icon = icon
+          minimalAsset.icons = icons
+        }
+
+        prev.byId[token.assetId] = makeAsset(state.assets.byId, minimalAsset)
+        prev.ids.push(token.assetId)
+
+        return prev
+      },
+      { byId: {}, ids: [] },
+    )
+  }
+
+  if (chainId === solanaChainId) {
+    const account = portfolioAccounts[pubkey] as Account<KnownChainIds.SolanaMainnet>
+
+    return (account.chainSpecific.tokens ?? []).reduce<UpsertAssetsPayload>(
+      (prev, token) => {
```

Comment:
```suggestion
      (prev, token) => {
      if (state.assets.byId[token.assetId]) return prev
```

Reply:
We display no token icons because of this missing line

Reply:
derp a3e0ef265b716b968c2fc5321b63b9b2923f3875